File: shared/src@jvm/io/github/yuanbaobaoo/petallink/ui/pages/main/Breadcrumb.kt
@file:Suppress("FunctionName")

package io.github.yuanbaobaoo.petallink.ui.pages.main

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.github.yuanbaobaoo.petallink.ui.components.mate.MateHDivider
import io.github.yuanbaobaoo.petallink.ui.theme.LOCAL_SEMANTIC_COLORS
import io.github.yuanbaobaoo.petallink.ui.theme.PetalTheme
import io.github.yuanbaobaoo.petallink.ui.viewmodel.BrowserBreadcrumb

/**
 * 面包屑导航（视觉对标 design/v2/02-main.html 的 .breadcrumb）。
 *
 * v2：高 40px，横向 scroll（超宽不换行），padding 0/20，gap 6；底部 MateHDivider 分隔线保留。
 * 分隔符 `›`（13sp placeholder 灰）；普通段 14sp secondary 可点 hover→PetalTheme.colors.brand；
 * 当前段 14sp primary + semibold + 不可点。
 *
 * @param crumbs 路径栈（最后一个为当前目录）
 * @param onNavigate 点击非末级段跳转
 */
@Composable
fun Breadcrumb(
    crumbs: List<BrowserBreadcrumb>,
    onNavigate: (BrowserBreadcrumb) -> Unit,
) {
    val semantic = LOCAL_SEMANTIC_COLORS.current
    val scroll = rememberScrollState()
    Column {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(PetalTheme.metrics.navigation.breadcrumbHeight)
                .background(semantic.bgContainer)
                .horizontalScroll(scroll)
                .padding(horizontal = PetalTheme.metrics.navigation.breadcrumbHorizontalPadding),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(PetalTheme.metrics.navigation.breadcrumbItemSpacing),
        ) {
            crumbs.forEachIndexed { index, crumb ->
                if (index > 0) {
                    Text("›", style = PetalTheme.typography.breadcrumb.separator, color = semantic.textPlaceholder)
                }
                val isCurrent = index == crumbs.lastIndex
                // hover 变 brand 色（对标 .breadcrumb__item:hover）
                val interactionSource = remember { MutableInteractionSource() }
                val hovered by interactionSource.collectIsHoveredAsState()
                val color = when {
                    isCurrent -> semantic.textPrimary
                    hovered -> PetalTheme.colors.brand
                    else -> semantic.textSecondary
                }
                Text(
                    crumb.name,
                    style = if (isCurrent) PetalTheme.typography.breadcrumb.currentItem else PetalTheme.typography.breadcrumb.item,
                    color = color,
                    modifier = Modifier
                        .hoverable(interactionSource = interactionSource, enabled = !isCurrent)
                        .then(
                            if (isCurrent) Modifier
                            else Modifier.clickable(
                                interactionSource = interactionSource,
                                indication = null,
                            ) { onNavigate(crumb) },
                        ),
                )
            }
        }
        // 底部分隔线（v2 保留 MateHDivider）
        MateHDivider()
    }
}

/**
 * 面包屑底部分隔线（保留兼容，实际已内嵌在 Breadcrumb 内）。
 */
@Composable
fun BreadcrumbDivider() {
    MateHDivider()
}

File: lib/pages/files/widgets/breadcrumb.dart
import 'package:flutter/material.dart';

import 'package:petal_link/app/theme/mate_theme.dart';
import 'package:petal_link/pages/files/controller/file_browser_controller.dart';
import 'package:petal_link/widgets/index.dart';

/// 面包屑导航（对标 CMP Breadcrumb.kt；v2：design/02-main.html .breadcrumb）。
///
/// 高 40px，横向 scroll（超宽不换行），padding 0/20，gap 6；底部 MateHDivider。
/// 分隔符 `›`（placeholder 灰）；普通段 secondary 可点 hover→brand；
/// 当前段 primary + semibold + 不可点。
class FilesBreadcrumb extends StatelessWidget {
  /// 路径栈（最后一个为当前目录）
  final List<Breadcrumb> crumbs;

  /// 点击非末级段跳转（传目标下标）
  final void Function(int index) onNavigate;

  const FilesBreadcrumb({
    super.key,
    required this.crumbs,
    required this.onNavigate,
  });

  @override
  Widget build(BuildContext context) {
    final colors = MateTheme.colorsOf(context);
    final typography = MateTheme.typographyOf(context);
    final metrics = MateTheme.metricsOf(context).navigation;

    return Column(
      children: [
        SizedBox(
          height: metrics.breadcrumbHeight,
          child: Container(
            color: colors.bgContainer,
            padding: EdgeInsets.symmetric(
              horizontal: metrics.breadcrumbHorizontalPadding,
            ),
            child: ListView.builder(
              scrollDirection: Axis.horizontal,
              itemCount: crumbs.length * 2 - 1,
              itemBuilder: (context, i) {
                if (i.isOdd) {
                  // 分隔符 ›（placeholder 灰）
                  return Padding(
                    padding: EdgeInsets.symmetric(
                      horizontal: metrics.breadcrumbItemSpacing / 2,
                    ),
                    child: Center(
                      child: Text(
                        '›',
                        style: typography.breadcrumb.separator.copyWith(
                          color: colors.textPlaceholder,
                        ),
                      ),
                    ),
                  );
                }
                final index = i ~/ 2;
                final crumb = crumbs[index];
                final isCurrent = index == crumbs.length - 1;
                return _BreadcrumbItem(
                  name: crumb.name,
                  isCurrent: isCurrent,
                  onTap: isCurrent ? null : () => onNavigate(index),
                );
              },
            ),
          ),
        ),
        // 底部分隔线（v2 保留 MateHDivider）
        const MateHDivider(),
      ],
    );
  }
}

/// 面包屑单个段（hover→brand 色，对齐 .breadcrumb__item:hover）。
class _BreadcrumbItem extends StatefulWidget {
  final String name;
  final bool isCurrent;
  final VoidCallback? onTap;

  const _BreadcrumbItem({
    required this.name,
    required this.isCurrent,
    this.onTap,
  });

  @override
  State<_BreadcrumbItem> createState() => _BreadcrumbItemState();
}

class _BreadcrumbItemState extends State<_BreadcrumbItem> {
  bool _hovered = false;

  @override
  Widget build(BuildContext context) {
    final colors = MateTheme.colorsOf(context);
    final typography = MateTheme.typographyOf(context);

    final color = widget.isCurrent
        ? colors.textPrimary
        : _hovered
            ? colors.brand
            : colors.textSecondary;

    return MouseRegion(
      cursor: widget.isCurrent
          ? SystemMouseCursors.basic
          : SystemMouseCursors.click,
      onEnter: (_) => setState(() => _hovered = true),
      onExit: (_) => setState(() => _hovered = false),
      child: GestureDetector(
        onTap: widget.onTap,
        child: Center(
          child: Text(
            widget.name,
            style: (widget.isCurrent
                    ? typography.breadcrumb.currentItem
                    : typography.breadcrumb.item)
                .copyWith(color: color),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
        ),
      ),
    );
  }
}
